plugins/obs-ffmpeg/vaapi-utils.c

```diff
@@ -0,0 +1,263 @@
+	default:
+		blog(LOG_ERROR,
+		     "VAAPI: Fail to get RC attribute from the %s %s of the device '%s'",
+		     vaProfileStr(profile), vaEntrypointStr(entrypoint),
+		     device_path);
+		return 0;
+	}
+}
+
+static bool vaapi_display_ep_combo_supported(VAProfile profile,
+					     VAEntrypoint entrypoint,
+					     VADisplay dpy,
+					     const char *device_path)
+{
+	uint32_t ret = vaapi_display_ep_combo_rate_controls(profile, entrypoint,
+							    dpy, device_path);
+	if (ret & VA_RC_CBR || ret & VA_RC_CQP || ret & VA_RC_VBR)
+		return true;
+
+	return false;
+}
+
+bool vaapi_device_rc_supported(VAProfile profile, VADisplay dpy, uint32_t rc,
+			       const char *device_path)
+{
+	uint32_t ret = vaapi_display_ep_combo_rate_controls(
+		profile, VAEntrypointEncSlice, dpy, device_path);
+	if (ret & rc)
+		return true;
+	ret = vaapi_display_ep_combo_rate_controls(
+		profile, VAEntrypointEncSliceLP, dpy, device_path);
+	if (ret & rc)
+		return true;
+
+	return false;
+}
+
+#define CHECK_PROFILE(ret, profile, va_dpy, device_path)                      \
+	if (vaapi_display_ep_combo_supported(profile, VAEntrypointEncSlice,   \
+					     va_dpy, device_path)) {          \
+		blog(LOG_DEBUG, "'%s' support encoding with %s", device_path, \
+		     vaProfileStr(profile));                                  \
+		ret |= true;                                                  \
+	}
+
+#define CHECK_PROFILE_LP(ret, profile, va_dpy, device_path)                   \
+	if (vaapi_display_ep_combo_supported(profile, VAEntrypointEncSliceLP, \
+					     va_dpy, device_path)) {          \
+		blog(LOG_DEBUG, "'%s' support low power encoding with %s",    \
+		     device_path, vaProfileStr(profile));                     \
+		ret |= true;                                                  \
+	}
+
+bool vaapi_display_h264_supported(VADisplay dpy, const char *device_path)
+{
+	bool ret = false;
+
+	CHECK_PROFILE(ret, VAProfileH264ConstrainedBaseline, dpy, device_path);
+	CHECK_PROFILE(ret, VAProfileH264Main, dpy, device_path);
+	CHECK_PROFILE(ret, VAProfileH264High, dpy, device_path);
+
+	if (!ret) {
+		CHECK_PROFILE_LP(ret, VAProfileH264ConstrainedBaseline, dpy,
+				 device_path);
+		CHECK_PROFILE_LP(ret, VAProfileH264Main, dpy, device_path);
+		CHECK_PROFILE_LP(ret, VAProfileH264High, dpy, device_path);
+	}
+
+	return ret;
+}
+
+bool vaapi_device_h264_supported(const char *device_path)
+{
+	bool ret = false;
+	VADisplay va_dpy;
+
+	va_dpy = vaapi_open_device(device_path, "vaapi_device_h264_supported");
+	if (!va_dpy)
+		return false;
+
+	ret = vaapi_display_h264_supported(va_dpy, device_path);
+
+	vaapi_close_device(va_dpy);
+
+	return ret;
+}
+
+bool vaapi_h264_supported(void)
+{
+	bool ret = false;
+
+	char path[32] = "/dev/dri/renderD1";
+	for (int i = 28;; i++) {
+		sprintf(path, "/dev/dri/renderD1%d", i);
+		if (access(path, F_OK) != 0)
+			break;
+
+		ret = vaapi_device_h264_supported(path);
+		if (ret) {
+			default_h264_device = strdup(path);
```

Comment:
this side effect is hard to see, personally id prefer you just move this code into `vaapi_get_h264_default_device` function and move `default_h264_device` into a function local static so you can return it immediately if its non-null. In  ffmpeg you can try and get the default device and if we return a null string we can assume no vaapi devices exist and return appropriately.

Reply:
Making `vaapi_get_h264_default_device` "local static" is not possible since I want to keep all pure VAAPI function in only one C file.

Reply:
Sorry im not sure what the appropriate name for what i mean is. https://gist.github.com/kkartaltepe/ce473226346f95c98470e5e9f5fd41a3 this is what i mean (but also just deleting the `vaapi_h264_supported` and calling `vaapi_get_h264_default_device` in the ffmpeg portion is probably even better).

Reply:
Made the change, to get rid of `vaapi_h264_supported` I added a comment to explain why `vaapi_get_h264_default_device` is used to check if H264 is supported.